macros/datediff.sql

```diff
@@ -0,0 +1,43 @@
+{%- macro datediff(part, left_val, right_val, date_format='%Y-%m-%d') -%}
+    {# determines the delta (in `part` units) between first_val and second_val.
+       *Note* the order of left_val, right_val is reversed from Snowflake.
+       ARGS:
+         - part (string) one of 'day', 'week', 'month', 'year'
+         - left_val (date/timestamp) the value before the minus in the equation "left - right"
+         - right_val (date/timestamp) the value after the minus in the equation "left - right"
+         - date_format (pattern) a string pattern for the provided arguments (primarily for BigQuery)
+       RETURNS: An integer representing the delta in `part` units
+    #}
+
+    {%if part not in ('day','week','month','year',) %}
+        {{exceptions.raise_compiler_error("macro datediff for target does not support date part value " ~ part)}}
+    {%- endif -%}
+
+    {%- if target.type ==  'postgres' -%} 
+        {%- if part == 'year' -%}
+            DATE_PART('year',{{left_val}}::DATE) - DATE_PART('year',{{right_val}}::DATE)
+        {%- elif part == 'month' -%}
+            ((DATE_PART('year',{{left_val}}::DATE) - DATE_PART('year',{{right_val}}::DATE)) * 12)
+            +
+            (DATE_PART('month',{{left_val}}::DATE) - DATE_PART('month',{{right_val}}::DATE))
+        {%- elif part == 'week' -%}
+            TRUNC(DATE_PART('day', ({{left_val}}::TIMESTAMP - {{right_val}}::TIMESTAMP))/7)
+        {%- elif part == 'day' -%}
+            DATE_PART('day', ({{left_val}}::TIMESTAMP - {{right_val}}::TIMESTAMP))
+        {%- endif %}
+
+    {%- elif target.type == 'bigquery' -%}
+
+        DATE_DIFF(PARSE_DATE('{{date_format}}', {{left_val}}), 
+                  PARSE_DATE('{{date_format}}', {{right_val}}), 
+                  {{part|upper}})
+
+    {%- elif target.type == 'snowflake' -%}
+
+        DATEDIFF('{{part|lower}}', {{right_val}},{{left_val}})
```

Comment:
If we're coercing to date / timestamp in pg, any value to doing it for bq / snowflake? Also, why not coerce them once (I'm assuming both also support timestamp / dates) to remove the need to constantly cast?

Reply:
so they don't all behave the same with coercing into timestamp / date, unfortnately :( the same TS values coerced the same between Snowflake and PG will give different diff values, which is gross. Bigquery does it's own weirdness as you can see with parsing first from strings. it behaves more like raw JS. 

however I think there is room to standardize here for sure. I will tackle in a new PR

Reply:
Lemme try to flatten this in the next PR when I add the h/m/s datediff. it'll be more obvious where we can dry it up in there I think.